samples/python/text_generation/react_sample.py

```diff
@@ -0,0 +1,215 @@
+                    tools_text=tools_text,
+                    tools_name_text=tools_name_text,
+                    query=query,
+                )
+        if query:
+            messages.append({"role": "user", "content": query})
+        if response:
+            messages.append({"role": "assistant", "content": response})
+
+    prompt = tokenizer.apply_chat_template(messages, add_generation_prompt=True)
+
+    return prompt
+
+
+def parse_latest_tool_call(text):
+    tool_name, tool_args = "", ""
+    i = text.rfind("\nAction:")
+    j = text.rfind("\nAction Input:")
+    k = text.rfind("\nObservation:")
+    if 0 <= i < j:  # If the text has `Action` and `Action input`,
+        if k < j:  # but does not contain `Observation`,
+            # then it is likely that `Observation` is ommited by the LLM,
+            # because the output text may have discarded the stop word.
+            text = text.rstrip() + "\nObservation:"  # Add it back.
+        k = text.rfind("\nObservation:")
+        tool_name = text[i + len("\nAction:") : j].strip()
+        tool_args = text[j + len("\nAction Input:") : k].strip()
+        text = text[:k]
+    return tool_name, tool_args, text
+
+
+def call_tool(tool_name: str, tool_args: str) -> str:
+    if tool_name == "get_weather":
+        try:
+            city_name = json5.loads(tool_args)["city_name"]
+        except KeyError:
+            print("Error: 'city_name' key not found in tool_args")  # Debug print
+            raise
+        key_selection = {
+            "current_condition": [
+                "temp_C",
+                "FeelsLikeC",
+                "humidity",
+                "weatherDesc",
+                "observation_time",
+            ],
+        }
+        resp = requests.get(f"https://wttr.in/{city_name}?format=j1")
+        resp.raise_for_status()
+        resp = resp.json()
+        ret = {k: {_v: resp[k][0][_v] for _v in v} for k, v in key_selection.items()}
+        return str(ret)
+    elif tool_name == "image_gen":
+        import urllib.parse
+
+        tool_args = tool_args.replace("(", "").replace(")", "")
+        prompt = json5.loads(tool_args)["prompt"]
+        prompt = urllib.parse.quote(prompt)
+        return json.dumps(
+            {"image_url": f"https://image.pollinations.ai/prompt/{prompt}"},
+            ensure_ascii=False,
+        )
+    else:
+        raise NotImplementedError
+
+
+def llm_with_tool(llm_pipe, llm_config, tokenizer, prompt: str, history, list_of_tool_info=()):
+    chat_history = [(x["user"], x["bot"]) for x in history] + [(prompt, "")]
+    planning_prompt = build_input_text(tokenizer, chat_history, list_of_tool_info)
+
+    text = ""
+    while True:
+        output = llm_pipe.generate(planning_prompt + text, llm_config, streamer)
+        action, action_input, output = parse_latest_tool_call(output)
+        if action:
+            print("Action:", action)
+            print("Action_input:", action_input)
```

Comment:
`Qwen2.5-3B-Instruct` generates two actions: `get_weather` and `image_gen` but only the last is printed. The whether is simply hallucinated: `Observation: {"city": "London", "country": "UK", "time": "2022-03-14 12:00:00", "temperature": "12°C", "condition": "Cloudy", "wind": "10 km/h"}`